### ISSUE TITLE
Add support for wrapping with text

### ISSUE THREAD

**Issue by hgiesel** [ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style#4]
I'm an avid user of your addon, and wondered whether you would add support for wrapping a selection with raw text, rather than a tag. An example:

```
Capital of Italy is Rome
```

You select the text "Rome". Then you press a shortcut defined in the Config dialog, and you end up with:

```
Capital of Italy is {{c1::Rome}}
```

Of course this would just recreate the cloze shortcut, however the difference would be that it is configurable.

My use case is specifically my add-on [Set Randomizer](https://ankiweb.net/shared/info/1164532380). I'm sure users would find other users.

I've been thinking about implementing it on my own for some time now, but it kinda exceeds the scope that I originally envisioned, and I think it would much better fit into this add-on.

I would write a pull request if it's okay with you.

**Comment by ijgnd** [comment]
Today I worked on "editor: apply font color, background color, custom class, custom style". One reason (besides using the js library rangy for applying classes) was that I wanted better night mode support so for each class you can now enter a regular style and a night mode style. Accordingly the main table in the settings window has two columns for this. Two columns works well for "text before" and "text after". 

So I also added an option for text wrapping, see [this commit](https://github.com/ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style/commit/04dde812689cacb9c2e57a3044ddf9b41ae0ab52). 

So you were right all along. Maybe it's still interesting for you?

If you are interested: You would have to check if special characters like backslashes are properly handled. I save the text from the guis to a python dict that I store in a pickle file. When inserting into the editor I use json.dumps(), see [here](https://github.com/ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style/blob/4bd1bb62bcabb68824823ce34271cc685e17c7eb/src/editor_apply_styling_functions.py#L39).

If you want to test it: an ankiaddon-file is [here](https://github.com/ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style/releases/tag/0.4). (only works on 2.1.22 or higher).

for building the add-on I use my own short build-script from [here](https://gist.github.com/ijgnd/17f3fdd2bf1e8457692f37b7ff01339a). This might be useful because 9 ui-files need to be processed for this add-on.

Applying the text wrapper option was quite easy after I had re-understood my original code. The code is still mostly unchanged because cleaning it up properly would have taken even longer - if you look at the commits history you'll see many changes from today but it's mostly cosmetic like more clear function names or file names. So I need to live with the limitations of my old config and that values are stored in keys with unexpected names ...

If you find bugs or have good ideas for the defaultconfig please let me know.

btw: Do you still have the problem described in #5?

**Comment by ijgnd** [comment]
I only tested on my main computer (linux) and just saw that my way of loading/initializing rangy doesn't work in windows. So maybe it also doesn't work on macos either or on other computers at all. So don't waste your time with the current version.

**Comment by hgiesel** [comment]
Hi,

I just tested it. It works... I think?

When creating I only have three options:
* Backcolor (via class)
* Forecolor
* class (other)
I cannot create any new Text wrappers.

However i noticed that there are already two present (which use some Set Randomizer syntax, haha).
I tested them and they work perfectly.

I'm not quite sure what you mean by "loading/initializing rangy".

**Comment by ijgnd** [comment]
tldr: new version to test is [here](https://github.com/ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style/releases/tag/0.5). Does this work for you? Thanks for testing & feedback.

&nbsp;

> When creating I only have three options:

Thanks. I forgot to update the list that's loaded into the listwidget where you can select which category to create. This was a quick fix.

&nbsp;

> rangy problems ...

"rangy" refers to this [js library](https://github.com/timdown/rangy) which among other things offers to apply classes to selected text. 

When I run Anki in Windows from the console with the add-on version from last night when I open a window with an editor I get these error messages:

```
JS error /_addons/1899278645/web/rangy-classapplier.js:15 Uncaught TypeError: Cannot read property 'createModule' of undefined
JS error /_addons/1899278645/web/rangy-core.js:10 Uncaught Error: required module 'ClassApplier' not found
JS error :41 Uncaught TypeError: rangy.createHighlighter is not a function
JS error :41 Uncaught TypeError: rangy.createHighlighter is not a function
JS error :41 Uncaught TypeError: rangy.createHighlighter is not a function
....
```
Last night I added the rangy js files into the webview with the 2.1.22-webview_will_set_content-hook, see [here](https://github.com/ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style/blob/4bd1bb62bcabb68824823ce34271cc685e17c7eb/src/editor_set_css_js_for_webview.py#L19). The js that prepares rangy was [here](https://github.com/ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style/blob/4bd1bb62bcabb68824823ce34271cc685e17c7eb/src/config.py#L42).

Because of my limited js knowledge today I decided that it's probably quicker to just work around the problem by extending [aqt.editor._html](https://github.com/ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style/blob/ddbb3b190dcb34a2af6726ae0a9c531f7e3176e0/src/config.py#L106). Side effect: It works in most old Anki 2.1 versions. If you have an idea all your fixes are still welcome.

&nbsp;

Today I also made two small fixes and I split up the file that held the code for GUIs into separate files. Some things are still missing for a finished version like a sensible default config, better updating of the templates, etc. (apart from the need of a general reorganization ...)

I'm also thinking about  extending the text wrapper function so that you can directly set the values for "before" and "after" for all entries in one separate dialog that avoids the main config dialog. Then it should be more useful for less permanent entries.

**Comment by ijgnd** [comment]
I felt bad about the solution from last night to load the js library. So I changed the code again. This time I'm using an async js function that's called from a python function that's called by a new style hook. So hopefully this is futureproof and works. Probably the last sentence was unclear. I'm referring to [this code](https://github.com/ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style/blob/9f7dd9498e3a021c4b9ca58ce31dd630c01cfd86/src/editor_set_css_js_for_webview.py#L52).

Maybe you can quickly see if this is a real solution or if this just works by accident ... I'll upload a compiled version later.

**Comment by hgiesel** [comment]
I'll be happy to test, but I'm afraid I also don't know about best practices for such a scenario. Just a few thoughts:

1. I remember trying to include some JavaScript into note templates before. I tried this using the collection folder.
2. As I see it, this new solution would only work on Desktop? Or am I wrong?

**Comment by ijgnd** [comment]
The new solution should still work on mobile. My changes only concerned how the custom css and js is loaded into the editor. I uploaded an ankiaddon file [here](https://github.com/ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style/releases/download/0.7/anki__editor__apply__font_color__background_color__custom_class__custom_style__2020-05-14.ankiaddon) of the version I talked about yesterday that contains some more enhancements and doesn't discourage you from using inline styling anymore. Maybe you could test if it works on your Mac? Thanks.
